ethstorage/prover/zk_prover_go.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2022-2023, EthStorage.
+// For license information, see https://github.com/ethstorage/es-node/blob/main/LICENSE
+
+package prover
+
+import (
+	"fmt"
+	"math/big"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/crate-crypto/go-proto-danksharding-crypto/eth"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/iden3/go-rapidsnark/prover"
+	"github.com/iden3/go-rapidsnark/witness/v2"
+	"github.com/iden3/go-rapidsnark/witness/wazero"
+)
+
+type ZKProverGo struct {
+	calc witness.Calculator
+	zkey []byte
+	lg   log.Logger
+}
+
+func NewZKProverGo(libDir, zkeyName, wasmName string, lg log.Logger) (*ZKProverGo, error) {
+	wasmBytes, err := os.ReadFile(filepath.Join(libDir, wasmName))
+	if err != nil {
+		lg.Error("Read wasm file failed", "error", err)
+		return nil, err
+	}
+	calc, err := witness.NewCalculator(wasmBytes, witness.WithWasmEngine(wazero.NewCircom2WZWitnessCalculator))
+	if err != nil {
+		lg.Error("Create witness calculator failed", "error", err)
+		return nil, err
+	}
+	zkey, err := os.ReadFile(filepath.Join(libDir, zkeyName))
+	if err != nil {
+		lg.Error("Read zkey file failed", "error", err)
+		return nil, err
+	}
+	return &ZKProverGo{
+		zkey: zkey,
+		calc: calc,
+		lg:   lg,
+	}, nil
+}
+
+// Generate ZK Proof for the given encoding key and sample index
+func (p *ZKProverGo) GenerateZKProof(encodingKeys []common.Hash, sampleIdxs []uint64) ([]byte, []*big.Int, error) {
+	proof, publics, err := p.GenerateZKProofRaw(encodingKeys, sampleIdxs)
+	if err != nil {
+		return nil, nil, err
+	}
+	if len(publics) != 6 {
+		return nil, nil, fmt.Errorf("publics length is %d", len(publics))
+	}
+	return proof, publics[4:], nil
+}
+
+func (p *ZKProverGo) GenerateZKProofRaw(encodingKeys []common.Hash, sampleIdxs []uint64) ([]byte, []*big.Int, error) {
+	for i, idx := range sampleIdxs {
+		p.lg.Debug("Generate zk proof", "encodingKey", encodingKeys[i], "sampleIdx", sampleIdxs[i])
+		if int(idx) >= eth.FieldElementsPerBlob {
```

Comment:
Maybe we can move the sampleIdx check into GenerateInputs func to reduce redundancy?